sdk/core/azure-core/inc/azure.hpp

```diff
@@ -3,12 +3,13 @@
 
 #pragma once
 
+#include <string>
 #include <internal/contract.hpp>
 
 #define AZURE_UNREFERENCED_PARAMETER(x) ((void) (x));
 
 namespace Azure { namespace Core {
 
-
+bool LocaleInvariantCaseInsensitiveEqual(const std::string& lhs, const std::string& rhs) noexcept;
```

Comment:
This looks good but I don't think we want this to be public. So maybe move to details namespace?

Reply:
Hmmm I would argue it should be public: there's no reason users won't want to mimic the same case comparison rules we use in our bits elsewhere in their application (e.g. they parse one of our enums out of one of their own config file formats or something like that). But this is more in "SDK Design" territory than "C++" territory so I would defer to you.

(Then again were it up to me none of this stuff would support case insensitivity :) )

Reply:
There's also a discussion point here: Are parts of Core that are "public" to other parts of the SDK "public" (i.e. is being in "core" a sufficient indication that it's support stuff?)

Reply:
"Details" is usually for things that aren't called outside that file at all

Reply:
We'd like to have functions in Core that are for our own internal implementation that are not public/supported so we can change them easily without breaking customers.

Reply:
Our SDK is about calling into Azure. We may need some helper functions (like case-insensitive comparison) for our own implementation but I don't want to expose any of our random helper functions as part of public/supported functionality. We don't want to doc, fully test, and support random helper functions.

Reply:
In our C SDK, we have the "private to core", "internal to core and other client libraries; but not to customers" and "public to core, client libs and customers". Some of this is accomplished via what directory a header file is in and some of this is accomplished via convention: functions/structs that start with an underscore are not public, fields in a nested _internal struct are not document and customers should avoid, etc. 

We need to come up with good conventions in C++ to accomplish similar doc/support boundaries. The Details namespace is one such convention. We should have a meeting to discuss the best approaches for various scenarios.